Fixes #28 & Fixes #40: No results issue

luan-cestari opened the pull request:
Cleaned up code (imports and etc) and created some new tests 

Related to: https://github.com/lightblue-platform/lightblue-client/issues/28

jewzaam commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -119,21 +127,30 @@ public LightblueResponse metadata(LightblueRequest lightblueRequest) {
+        } catch (JsonMappingException e) {
+            LOGGER.error("Error parsing lightblue response: " + responseJson.toString(), e);
+            throw new RuntimeException("Error parsing lightblue response: " + responseJson.toString());
         }
```
This change in data(LightblueRequest, Class<T>) reverts changes to use new LightblueResponse#parseProceseed method.  Why?

luan-cestari replied:
I will revert/fix this

jewzaam commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -119,21 +127,30 @@ public LightblueResponse metadata(LightblueRequest lightblueRequest) {
-            throw new LightblueHttpClientException("Error sending lightblue request: " + lightblueRequest.getBody(), e);
-        }
+        return callService(new LightblueHttpDataRequest(lightblueRequest)
+                .getRestRequest(configuration.getDataServiceURI()));
```
Method data(LightblueRequest) no longer wraps runtime exception.  I can see why that, at a glance, would seem bad.  Just want to be sure this change doesn't have unexpected side effects.  the difference is the request JSON won't be part of the stacktrace anymore.  I have mixed feelings about this.  1) it would be easier to debug yet 2) it means possible sensitive data is logged on the client.

luan-cestari replied:
Well,  I think I can change it back how it was and re-throw  the exception in data(LightblueRequest lightblueRequest, Class<T> type)  method, what do you think?

jewzaam replied:
Was the change you made to fix the issue you're trying to fix?  I don't know what is changing in this PR to solve that issue, though I need to do another pass through the changes sometime.

jewzaam commented on http/src/test/java/com/redhat/lightblue/client/http/LightblueHttpClientTest.java:
```diff
@@ -22,29 +26,93 @@ public void testPojoMapping() throws IOException {
+    }
+
+    @Test
+    public void testPojoMappingWithEmptyProcessedResults_ForSingleResult() throws IOException {
```
testPojoMappingWithEmptyProcessedResults_ForSingleResult name is misleading, maybe call it testPojoMappingWithEmptyProcessedResults_ForEmptyJsonNode ?  I see in the test it's verifying the client gracefully handles parsing an empty array to a single null result.

luan-cestari replied:
I will update with your suggestion

luan-cestari commented:
Updated the PR. Now almost an LightblueHttpClientException instance is used to wrap the exceptions we caught (only IOException instances are ignored as it is part of the method signature). Added also a line commented why some places the try/catch block was removed as it turn to be redundant for the current implementation

luan-cestari commented:
Noticed that many classes is missing Copyright header, may I create a PR for this?

dcrissman commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -129,11 +131,25 @@ public LightblueResponse data(LightblueRequest lightblueRequest) {
+        } catch (JsonParseException | JsonMappingException e) {
+            LOGGER.error("Error parsing lightblue response: " + responseJson.toString(), e);
+            throw new LightblueHttpClientException("Error parsing lightblue response: " + responseJson.toString());
         }
```
It looks like you are reverting changes made in #64

dcrissman commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -157,12 +173,12 @@ protected LightblueResponse callService(HttpRequestBase httpOperation) {
                     LOGGER.debug("Response received from service" + jsonOut);
-                    return new LightblueResponse(jsonOut, mapper);
+                    return new LightblueResponse(jsonOut);
                 }
```
What if a consumer of LightblueHttpClient wants to set a ObjectMapper, how would the LightblueResponse know about it?

dcrissman commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -157,12 +173,12 @@ protected LightblueResponse callService(HttpRequestBase httpOperation) {
             LOGGER.error("There was a problem calling the lightblue service", e);
-            return new LightblueResponse("{\"error\":\"There was a problem calling the lightblue service\"}", mapper);
+            return new LightblueResponse("{\"error\":\"There was a problem calling the lightblue service\"}");
         }
```
Same question as above.

dcrissman commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -129,11 +131,25 @@ public LightblueResponse data(LightblueRequest lightblueRequest) {
+        if (responseNode instanceof MissingNode || (responseNode instanceof ArrayNode && responseNode.size() == 0 && !type.isArray())) {
+            return null;
+        }
+
```
Isn't this being handled in LightblueResponse#parseProcessed?

dcrissman commented:
@luan-cestari I think I see the redundancy you referenced, I submitted #72 to fix. Good catch, thanks.

luan-cestari commented:
@dcrissman thank for reviewing and nice comments. I just pushed an update. ALso, What do you think about the Copyright thing?

dcrissman commented:
@luan-cestari I agree about the copywrite?

dcrissman commented on http/src/main/java/com/redhat/lightblue/client/http/LightblueHttpClient.java:
```diff
@@ -55,18 +70,15 @@ public LightblueHttpClient(String configFilePath) {
     public LightblueHttpClient(LightblueClientConfiguration configuration) {
-        this(configuration, LightblueResponse.DEFAULT_MAPPER);
+        this(configuration, DEFAULT_MAPPER);
     }
```
In #64 I moved this constant to LightblueResponse thinking that we could have a single default mapper defined and use it rather than individual classes defining their own. Maybe a more centralized location would be better? Thoughts?

dcrissman replied:
This is discussed further in #65

luan-cestari replied:
I will revert to previous way, so the PR for the #65 will fix this, ok?

luan-cestari commented:
Updated the PR

dcrissman commented:
@luan-cestari Before I merge this, I am fuzzy on what specifically this PR is intended to do. I see a lot of clean up, but would you mind updating the description to include a summary?

dcrissman commented:
@luan-cestari Could you also merge master in?

luan-cestari commented:
@dcrissman Updated the issue description and merged with master.

dcrissman commented on core/src/main/java/com/redhat/lightblue/client/response/LightblueResponse.java:
```diff
@@ -101,6 +101,11 @@ private int parseInt(String fieldName) {
+            if (errorCode != null && !errorCode.isNull() && !errorCode.isMissingNode()){
+                throw new LightblueResponseParseException("Error parsing lightblue response: Error returned. JSON returned:" + json.toString() + "\n");
+            }
+
```
How is this different, or could it be included in hasError?

luan-cestari replied:
I got some NPE in the tests and the exception was pretty similar something I already saw on ligthblue itself, but I confess I didn't look the implementation of has Error , it might be better if there is something missing to put on that method. I will check

luan-cestari replied:
I will try to make a fix o the hasError method, which would include to check errorCode in the json response and I also thought it would be nice to check the HTTP status codes (mainly because if we got the deployment not running and the application server or even the load balancer could return an error and that would not be noticed but that method right now )

luan-cestari commented:
Updated the PR so hasError could be used stead leaving an error validation somewhere else

alechenninger commented:
Is there a code style / import order guide for lightblue projects? (re: code cleanup in this PR)

jewzaam commented:
@dcrissman is this ready to merge?  I see you were reviewing..

@alechenninger http://dev.docs.lightblue.io/formatting/README.html is as close as we have

dcrissman commented:
My vote is to close this PR.

jewzaam commented:
I took a look through it and I *think* there are some small changes we might look at cherry picking.  But this PR needs to be closed, the changes if we want them should be implemented in a new PR.

LightblueHttpDataRequest
LightblueHttpMetadataRequest
LightblueHttpClientTest
LightblueResponse
- hasError

dcrissman commented:
So  much has changed in the client since this PR was created, I don't think there is anything of value here any more. I am going to close it out.
